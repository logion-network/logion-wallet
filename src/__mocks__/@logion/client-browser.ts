export class BrowserFile {

    constructor(blob: Blob) {
        this.blob = blob;
    }

    private blob: Blob;

    getBlob() {
        return this.blob;
    }
}

export class BrowserAxiosFileUploader {

    async upload() {

    }
}
