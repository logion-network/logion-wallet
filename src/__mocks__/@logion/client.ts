import { CheckResultType } from "@logion/client/dist/CollectionItem.js";
import { DefaultSignAndSendStrategy } from "@logion/client/dist/Signer.js";

import {
    LogionClient,
    AccountTokens,
    toIsoString,
    fromIsoString,
    ClosedCollectionLoc,
    DraftRequest,
    EditableRequest,
    OpenLoc,
    LocRequestState,
    LocsState,
    HashOrContent,
} from '../LogionClientMock';

export {
    LogionClient,
    AccountTokens,
    toIsoString,
    fromIsoString,
    CheckResultType,
    ClosedCollectionLoc,
    DraftRequest,
    EditableRequest,
    OpenLoc,
    DefaultSignAndSendStrategy,
    LocRequestState,
    LocsState,
    HashOrContent,
}

export { MimeType } from "@logion/client/dist/Mime.js";
