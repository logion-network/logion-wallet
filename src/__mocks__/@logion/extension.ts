import {
    enableExtensions,
    ExtensionSigner,
    isExtensionAvailable,
    enableMetaMask,
    allMetamaskAccounts,
} from '../PolkadotExtensionDappMock';
export {
    enableExtensions,
    ExtensionSigner,
    isExtensionAvailable,
    enableMetaMask,
    allMetamaskAccounts,
};
