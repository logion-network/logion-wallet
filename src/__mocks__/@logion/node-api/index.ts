import { ApiPromise } from "src/__mocks__/PolkadotApiMock";
export * from "@logion/node-api/dist/UUID";

export function buildApi() {
    return new ApiPromise();
}

export * from "./dist/LogionLoc";
