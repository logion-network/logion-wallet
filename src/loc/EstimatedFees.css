.EstimatedFees table {
    margin-left: auto;
    margin-right: auto;
}

.EstimatedFees td:first-child {
    font-weight: bold;
    text-align: left;
}

.EstimatedFees td:nth-child(2) {
    text-align: right;
    width: 200px;
}
