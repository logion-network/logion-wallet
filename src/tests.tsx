import { ReactElement } from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import reactRenderer, { ReactTestRenderer, ReactTestRendererJSON, act as reactAct } from 'react-test-renderer';
import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

export function render(element: ReactElement): null | ReactTestRendererJSON | ReactTestRendererJSON[] {
    const component = reactRenderer.create(element);
    return component.toJSON();
}

export function renderer(element: ReactElement): ReactTestRenderer {
    return reactRenderer.create(element);
}

export function shallowRender<E extends ReactElement>(element: ReactElement): E {
    const renderer = createRenderer();
    renderer.render(element);
    return renderer.getRenderOutput();
}

export function mockAccount(address: string, name: string) {
    return {
        address,
        meta: {
            name
        }
    };
}

export { reactAct as act };

export async function clickByName(name: string | RegExp | ((accessibleName: string, element: Element) => boolean)) {
    let button: HTMLElement;
    await waitFor(() => button = screen.getByRole("button", { name }));
    await userEvent.click(button!);
}

export async function typeByLabel(label: string, value: string) {
    let input: HTMLElement;
    await waitFor(() => input = screen.getByLabelText(label));
    await userEvent.type(input!, value);
}
