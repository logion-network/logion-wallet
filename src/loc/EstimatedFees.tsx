import { Fees, Numbers, Currency } from "@logion/node-api";
import AmountFormat from "src/common/AmountFormat";
import { customClassName } from "src/common/types/Helpers";
import "./EstimatedFees.css";

export interface Props {
    fees: Fees | undefined | null;
    hideTitle?: boolean;
    centered?: boolean;
    inclusionFeePaidBy?: string;
    otherFeesPaidBy?: string;
}

export default function EstimatedFees(props: Props) {
    if(!props.fees) {
        return null;
    }

    const className = customClassName("EstimatedFees", props.centered === true ? "centered-table" : undefined);
    return (
        <div className={ className }>
            { props.hideTitle !== true && <h3>Estimated fees (LGNT)</h3> }
            <table>
                <tbody>
                    <tr>
                        <td>Blockchain record</td>
                        <td>
                            <AmountFormat
                                amount={ Currency.toPrefixedNumberAmount(props.fees.inclusionFee).convertTo(Numbers.NONE) }
                                decimals={4}
                            />
                        </td>
                        { (props.otherFeesPaidBy !== undefined || props.inclusionFeePaidBy !== undefined) && <td>{ props.inclusionFeePaidBy || "" }</td> }
                    </tr>
                    { props.fees.storageFee !== undefined && 
                        <tr>
                            <td>File storage</td>
                            <td>
                                <AmountFormat
                                    amount={ Currency.toPrefixedNumberAmount(props.fees.storageFee).convertTo(Numbers.NONE) }
                                    decimals={4}
                                />
                            </td>
                            { (props.otherFeesPaidBy !== undefined || props.inclusionFeePaidBy !== undefined) && <td>{ props.otherFeesPaidBy || "" }</td> }
                        </tr>
                    }
                    { props.fees.legalFee !== undefined && 
                        <tr>
                            <td>Legal fee</td>
                            <td>
                                <AmountFormat
                                    amount={ Currency.toPrefixedNumberAmount(props.fees.legalFee).convertTo(Numbers.NONE) }
                                    decimals={4}
                                />
                            </td>
                            { (props.otherFeesPaidBy !== undefined || props.inclusionFeePaidBy !== undefined) && <td>{ props.otherFeesPaidBy || "" }</td> }
                        </tr>
                    }
                    <tr>
                        <td>Total</td>
                        <td>
                            <AmountFormat
                                amount={ Currency.toPrefixedNumberAmount(props.fees.totalFee).convertTo(Numbers.NONE) }
                                decimals={4}
                            />
                        </td>
                        { (props.otherFeesPaidBy !== undefined || props.inclusionFeePaidBy !== undefined) && <td></td> }
                    </tr>
                </tbody>
            </table>
        </div>
    );
}
