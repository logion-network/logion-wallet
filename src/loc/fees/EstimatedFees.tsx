import { LocData } from "@logion/client";
import { Fees, Numbers, Currency } from "@logion/node-api";
import AmountFormat from "../../common/AmountFormat";
import { customClassName } from "../../common/types/Helpers";
import "../EstimatedFees.css";
import { useMemo } from "react";
import { LocItem } from "../LocItem";

export interface Props {
    fees: Fees | undefined | null;
    hideTitle?: boolean;
    centered?: boolean;
    inclusionFeePaidBy?: string;
    otherFeesPaidBy?: string;
}

export default function EstimatedFees(props: Props) {

    const showPaidBy = useMemo(() => {
        return props.otherFeesPaidBy !== undefined || props.inclusionFeePaidBy !== undefined;
    }, [ props.otherFeesPaidBy, props.inclusionFeePaidBy ]);

    if(!props.fees) {
        return null;
    }

    const className = customClassName("EstimatedFees", props.centered === true ? "centered-table" : undefined);
    return (
        <div className={ className }>
            { props.hideTitle !== true && <h3>Estimated fees (LGNT)</h3> }
            <table>
                <tbody>
                    <tr>
                        <td>Blockchain record</td>
                        <td>
                            <AmountFormat
                                amount={ Currency.toPrefixedNumberAmount(props.fees.inclusionFee).convertTo(Numbers.NONE) }
                                decimals={4}
                            />
                        </td>
                        { showPaidBy && <td>{ props.inclusionFeePaidBy || "" }</td> }
                    </tr>
                    { props.fees.storageFee !== undefined && 
                        <tr>
                            <td>File storage</td>
                            <td>
                                <AmountFormat
                                    amount={ Currency.toPrefixedNumberAmount(props.fees.storageFee).convertTo(Numbers.NONE) }
                                    decimals={4}
                                />
                            </td>
                            { showPaidBy && <td>{ props.otherFeesPaidBy || "" }</td> }
                        </tr>
                    }
                    { props.fees.legalFee !== undefined && 
                        <tr>
                            <td>Legal fee</td>
                            <td>
                                <AmountFormat
                                    amount={ Currency.toPrefixedNumberAmount(props.fees.legalFee).convertTo(Numbers.NONE) }
                                    decimals={4}
                                />
                            </td>
                            { showPaidBy && <td>{ props.otherFeesPaidBy || "" }</td> }
                        </tr>
                    }
                    { props.fees.certificateFee !== undefined && 
                        <tr>
                            <td>Certificate fee</td>
                            <td>
                                <AmountFormat
                                    amount={ Currency.toPrefixedNumberAmount(props.fees.certificateFee).convertTo(Numbers.NONE) }
                                    decimals={4}
                                />
                            </td>
                            { showPaidBy && <td>{ props.otherFeesPaidBy || "" }</td> }
                        </tr>
                    }
                    { props.fees.valueFee !== undefined && 
                        <tr>
                            <td>Value fee</td>
                            <td>
                                <AmountFormat
                                    amount={ Currency.toPrefixedNumberAmount(props.fees.valueFee).convertTo(Numbers.NONE) }
                                    decimals={4}
                                />
                            </td>
                            { showPaidBy && <td>{ props.otherFeesPaidBy || "" }</td> }
                        </tr>
                    }
                    <tr>
                        <td>Total</td>
                        <td>
                            <AmountFormat
                                amount={ Currency.toPrefixedNumberAmount(props.fees.totalFee).convertTo(Numbers.NONE) }
                                decimals={4}
                            />
                        </td>
                        { showPaidBy && <td></td> }
                    </tr>
                </tbody>
            </table>
        </div>
    );
}

export function getOtherFeesPaidBy(loc: LocData) {
    if(loc.requesterLocId === undefined && loc.sponsorshipId === undefined) {
        return PAID_BY_REQUESTER;
    } else if(loc.sponsorshipId) {
        return PAID_BY_SPONSOR;
    } else {
        return PAID_BY_LEGAL_OFFICER;
    }
}

export const PAID_BY_REQUESTER = "paid by requester";
export const PAID_BY_SPONSOR = "paid by sponsor";
export const PAID_BY_LEGAL_OFFICER = "paid by legal officer";

export function geInclusionFeePaidBy(loc: LocData, item: LocItem) {
    return item.submitter?.type === "Polkadot" && item.submitter?.address === loc.ownerAddress ? PAID_BY_LEGAL_OFFICER : PAID_BY_SUBMITTER;
}

export const PAID_BY_SUBMITTER = "paid by submitter";
