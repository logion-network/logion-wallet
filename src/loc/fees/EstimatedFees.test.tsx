import { Fees } from "@logion/node-api";
import { shallowRender } from "../../tests";
import EstimatedFees from "./EstimatedFees";
import { Lgnt } from "@logion/node-api";

jest.unmock("@logion/client");
jest.unmock("@logion/node-api");

describe("EstimatedFees", () => {

    it("renders fees without storage", () => {
        const fees = new Fees({ inclusionFee: Lgnt.fromCanonical(42n) });
        const element = shallowRender(<EstimatedFees fees={ fees } />);
        expect(element).toMatchSnapshot();
    });

    it("renders fees with storage", () => {
        const fees = new Fees({
            inclusionFee: Lgnt.fromCanonical(42n),
            storageFee: Lgnt.fromCanonical(32n),
        });
        const element = shallowRender(<EstimatedFees fees={ fees } />);
        expect(element).toMatchSnapshot();
    });

    it("renders fees with storage and paid by", () => {
        const fees = new Fees({
            inclusionFee: Lgnt.fromCanonical(42n),
            storageFee: Lgnt.fromCanonical(32n),
        });
        const element = shallowRender(<EstimatedFees
            fees={ fees }
            inclusionFeePaidBy="paid by the Legal Officer"
            otherFeesPaidBy="paid by the requester"
        />);
        expect(element).toMatchSnapshot();
    });

    it("renders fees with legal", () => {
        const fees = new Fees({
            inclusionFee: Lgnt.fromCanonical(42n),
            legalFee: Lgnt.fromCanonical(22n),
        });
        const element = shallowRender(<EstimatedFees fees={ fees } />);
        expect(element).toMatchSnapshot();
    });

    it("renders fees with certificate", () => {
        const fees = new Fees({
            inclusionFee: Lgnt.fromCanonical(42n),
            certificateFee: Lgnt.fromCanonical(22n),
        });
        const element = shallowRender(<EstimatedFees fees={ fees } />);
        expect(element).toMatchSnapshot();
    });
});
