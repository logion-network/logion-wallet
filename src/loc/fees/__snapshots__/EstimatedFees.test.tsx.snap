// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EstimatedFees renders fees with certificate 1`] = `
<div
  className="EstimatedFees"
>
  <h3>
    Estimated fees (LGNT)
  </h3>
  <table>
    <tbody>
      <tr>
        <td>
          Blockchain record
        </td>
        <td>
          <AmountFormat
            amount={
              Lgnt {
                "canonical": 42n,
              }
            }
            decimals={4}
            unit={
              Object {
                "symbol": "",
                "tenExponent": 0,
              }
            }
          />
        </td>
      </tr>
      <tr>
        <td>
          Certificate fee
        </td>
        <td>
          <AmountFormat
            amount={
              Lgnt {
                "canonical": 22n,
              }
            }
            decimals={4}
            unit={
              Object {
                "symbol": "",
                "tenExponent": 0,
              }
            }
          />
        </td>
      </tr>
      <tr>
        <td>
          Total
        </td>
        <td>
          <AmountFormat
            amount={
              Lgnt {
                "canonical": 64n,
              }
            }
            decimals={4}
            unit={
              Object {
                "symbol": "",
                "tenExponent": 0,
              }
            }
          />
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`EstimatedFees renders fees with legal 1`] = `
<div
  className="EstimatedFees"
>
  <h3>
    Estimated fees (LGNT)
  </h3>
  <table>
    <tbody>
      <tr>
        <td>
          Blockchain record
        </td>
        <td>
          <AmountFormat
            amount={
              Lgnt {
                "canonical": 42n,
              }
            }
            decimals={4}
            unit={
              Object {
                "symbol": "",
                "tenExponent": 0,
              }
            }
          />
        </td>
      </tr>
      <tr>
        <td>
          Legal fee
        </td>
        <td>
          <AmountFormat
            amount={
              Lgnt {
                "canonical": 22n,
              }
            }
            decimals={4}
            unit={
              Object {
                "symbol": "",
                "tenExponent": 0,
              }
            }
          />
        </td>
      </tr>
      <tr>
        <td>
          Total
        </td>
        <td>
          <AmountFormat
            amount={
              Lgnt {
                "canonical": 64n,
              }
            }
            decimals={4}
            unit={
              Object {
                "symbol": "",
                "tenExponent": 0,
              }
            }
          />
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`EstimatedFees renders fees with storage 1`] = `
<div
  className="EstimatedFees"
>
  <h3>
    Estimated fees (LGNT)
  </h3>
  <table>
    <tbody>
      <tr>
        <td>
          Blockchain record
        </td>
        <td>
          <AmountFormat
            amount={
              Lgnt {
                "canonical": 42n,
              }
            }
            decimals={4}
            unit={
              Object {
                "symbol": "",
                "tenExponent": 0,
              }
            }
          />
        </td>
      </tr>
      <tr>
        <td>
          File storage
        </td>
        <td>
          <AmountFormat
            amount={
              Lgnt {
                "canonical": 32n,
              }
            }
            decimals={4}
            unit={
              Object {
                "symbol": "",
                "tenExponent": 0,
              }
            }
          />
        </td>
      </tr>
      <tr>
        <td>
          Total
        </td>
        <td>
          <AmountFormat
            amount={
              Lgnt {
                "canonical": 74n,
              }
            }
            decimals={4}
            unit={
              Object {
                "symbol": "",
                "tenExponent": 0,
              }
            }
          />
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`EstimatedFees renders fees with storage and paid by 1`] = `
<div
  className="EstimatedFees"
>
  <h3>
    Estimated fees (LGNT)
  </h3>
  <table>
    <tbody>
      <tr>
        <td>
          Blockchain record
        </td>
        <td>
          <AmountFormat
            amount={
              Lgnt {
                "canonical": 42n,
              }
            }
            decimals={4}
            unit={
              Object {
                "symbol": "",
                "tenExponent": 0,
              }
            }
          />
        </td>
        <td>
          paid by the Legal Officer
        </td>
      </tr>
      <tr>
        <td>
          File storage
        </td>
        <td>
          <AmountFormat
            amount={
              Lgnt {
                "canonical": 32n,
              }
            }
            decimals={4}
            unit={
              Object {
                "symbol": "",
                "tenExponent": 0,
              }
            }
          />
        </td>
        <td>
          paid by the requester
        </td>
      </tr>
      <tr>
        <td>
          Total
        </td>
        <td>
          <AmountFormat
            amount={
              Lgnt {
                "canonical": 74n,
              }
            }
            decimals={4}
            unit={
              Object {
                "symbol": "",
                "tenExponent": 0,
              }
            }
          />
        </td>
        <td />
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`EstimatedFees renders fees without storage 1`] = `
<div
  className="EstimatedFees"
>
  <h3>
    Estimated fees (LGNT)
  </h3>
  <table>
    <tbody>
      <tr>
        <td>
          Blockchain record
        </td>
        <td>
          <AmountFormat
            amount={
              Lgnt {
                "canonical": 42n,
              }
            }
            decimals={4}
            unit={
              Object {
                "symbol": "",
                "tenExponent": 0,
              }
            }
          />
        </td>
      </tr>
      <tr>
        <td>
          Total
        </td>
        <td>
          <AmountFormat
            amount={
              Lgnt {
                "canonical": 42n,
              }
            }
            decimals={4}
            unit={
              Object {
                "symbol": "",
                "tenExponent": 0,
              }
            }
          />
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
